subxt/src/extrinsics/extrinsics_type.rs

```diff
@@ -0,0 +1,344 @@
+    address_start_idx: usize,
+    /// The end index of the address in the encoded `bytes`.
+    address_end_idx: usize,
+    /// The start index in the `bytes` from which the call is encoded.
+    call_start_idx: usize,
+    /// The pallet index.
+    pallet_index: u8,
+    /// The variant index.
+    variant_index: u8,
+    /// Extrinsic bytes.
+    bytes: Arc<[u8]>,
+}
+
+impl ExtrinsicDetails {
+    // Attempt to dynamically decode a single event from our events input.
+    fn decode_from(
+        metadata: Metadata,
+        extrinsic_bytes: Arc<[u8]>,
+        ids: ExtrinsicIds,
+        index: usize,
+    ) -> Result<ExtrinsicDetails, Error> {
+        const SIGNATURE_MASK: u8 = 0b1000_0000;
+        const VERSION_MASK: u8 = 0b0111_1111;
+        const LATEST_EXTRINSIC_VERSION: u8 = 4;
+
+        // Extrinsic are encoded in memory in the following way:
+        //   - first byte: abbbbbbb (a = 0 for unsigned, 1 for signed, b = version)
+        //   - signature: [unknown TBD with metadata].
+        //   - extrinsic data
+        if extrinsic_bytes.is_empty() {
+            return Err(ExtrinsicError::InsufficientData.into());
+        }
+
+        let version = extrinsic_bytes[0] & VERSION_MASK;
+        if version != LATEST_EXTRINSIC_VERSION {
+            return Err(ExtrinsicError::UnsupportedVersion(version).into());
+        }
+
+        let is_signed = extrinsic_bytes[0] & SIGNATURE_MASK != 0;
+
+        // Skip over the first byte which denotes the version and signing.
+        let cursor = &mut &extrinsic_bytes[1..];
+
+        let mut address_start_idx = 0;
+        let mut address_end_idx = 0;
+
+        if is_signed {
+            address_start_idx = extrinsic_bytes.len() - cursor.len();
+
+            // Skip over the address, signature and extra fields.
+            scale_decode::visitor::decode_with_visitor(
+                cursor,
+                ids.address,
+                &metadata.runtime_metadata().types,
+                scale_decode::visitor::IgnoreVisitor,
+            )
+            .map_err(scale_decode::Error::from)?;
+            address_end_idx = extrinsic_bytes.len() - cursor.len();
+
+            scale_decode::visitor::decode_with_visitor(
+                cursor,
+                ids.signature,
+                &metadata.runtime_metadata().types,
+                scale_decode::visitor::IgnoreVisitor,
+            )
+            .map_err(scale_decode::Error::from)?;
+
+            scale_decode::visitor::decode_with_visitor(
+                cursor,
+                ids.extra,
+                &metadata.runtime_metadata().types,
+                scale_decode::visitor::IgnoreVisitor,
+            )
+            .map_err(scale_decode::Error::from)?;
+        }
+
+        let call_start_idx = extrinsic_bytes.len() - cursor.len();
+
+        // Ensure the provided bytes are sound.
+        scale_decode::visitor::decode_with_visitor(
+            cursor,
```

Comment:
Nit: instead of moving the cursor here and resetting it below, perhaps here use `&mut *cursor` to "copy" the cursor and then we don't need to re-declare it below?